src/traces/scattermapbox/plot.js

```diff
@@ -0,0 +1,130 @@
+/**
+* Copyright 2012-2016, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var convert = require('./convert');
+
+
+function ScatterMapbox(mapbox, uid) {
+    this.mapbox = mapbox;
+    this.map = mapbox.map;
+
+    this.uid = uid;
+
+    this.idSourceFill = uid + '-source-fill';
+    this.idSourceLine = uid + '-source-line';
+    this.idSourceCircle = uid + '-source-circle';
+    this.idSourceSymbol = uid + '-source-symbol';
+
+    this.idLayerFill = uid + '-layer-fill';
+    this.idLayerLine = uid + '-layer-line';
+    this.idLayerCircle = uid + '-layer-circle';
+    this.idLayerSymbol = uid + '-layer-symbol';
```

Comment:
a `scattermapbox` trace is in fact 4 `mapbox-gl` sources and 4 layers. Each layer's visibility is determined in the `convert` step.